models/transformer/decoder_impl.hpp

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file models/transformer/decoder_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Transformer Decoder class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_TRANSFORMER_DECODER_IMPL_HPP
+#define MODELS_TRANSFORMER_DECODER_IMPL_HPP
+
+#include "decoder.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename Memory, class ActivationFunction, typename RegularizerType,
+    typename InputDataType, typename OutputDataType>
+TransformerDecoder<Memory, ActivationFunction, RegularizerType, InputDataType,
+OutputDataType>::TransformerDecoder() :
+    tgtSeqLen(0),
+    srcSeqLen(0),
+    memoryModule(NULL),
+    dModel(0),
+    numHeads(0),
+    dimFFN(0),
+    dropout(0)
+{
+  // Nothing to do here.
+}
+
+template<typename Memory, class ActivationFunction, typename RegularizerType,
+    typename InputDataType, typename OutputDataType>
+TransformerDecoder<Memory, ActivationFunction, RegularizerType, InputDataType,
+OutputDataType>::TransformerDecoder(const size_t tgtSeqLen,
+    const size_t srcSeqLen,
+    const Memory& memoryModule,
+    const size_t dModel,
+    const size_t numHeads,
+    const size_t dimFFN,
+    const double dropout) :
+    tgtSeqLen(tgtSeqLen),
+    srcSeqLen(srcSeqLen),
+    memoryModule(memoryModule),
+    dModel(dModel),
+    numHeads(numHeads),
+    dimFFN(dimFFN),
+    dropout(dropout)
+{
+  Concat<> concatQuery;
+  concatQuery.Add<IdentityLayer<>>();
+  concatQuery.Add<IdentityLayer<>>();
+  concatQuery.Add<IdentityLayer<>>();
+
+  attnBlock1.Add(concatQuery);
+  attnBlock1.Add<MultiheadAttention<ActivationFunction, RegularizerType,
+      InputDataType, OutputDataType> >(tgtSeqLen, srcSeqLen, dModel, numHeads);
+  attnBlock1.Add<Dropout<InputDataType, OutputDataType> >(dropout);
+
+  network.Add(attnBlock1);
+  network.Add<LayerNorm<InputDataType, OutputDataType> >(dModel * tgtSeqLen);
+
+  // if (key.is_empty())
+  //   Log::Fatal << "The 'key' matrix is required but empty!" << std::endl;
+
+  // MultiheadAttention<InputDataType, OutputDataType, RegularizerType>
+  //     encoderDecoderAttn(dModel, numHeads);
+  // encoderDecoderAttn.Key() = key;
```

Comment:
Regarding your question. If I'm not mistaken, you need to concatenate the output of the Encoder with the previous layer and pass it to `encoderDecoderAttn`. But don't forget the Encoder and the Decoder have different inputs. You need to split a dataset record into two parts: `(input, target)`. You can use the `Subview` layer for that. I tried to implement a rough model [here](https://github.com/mlpack/mlpack/pull/2375#issuecomment-672910701). I think it should provide a clue on your question.

Tell me if you have any questions. It's not that easy.

Reply:
The problem is neither we can pass more than one inputs to `FFN` nor to the `Concat` layer. I can understand the use of Concat for repeating query into `[query query query]` but how will we use it to get `[query key key]` or 
`[query key value]`? I think defining the `Backward` and `Gradient` function manually is not a bad option (It will use less number of layers also). Like if we define single encoder and single decoder block as a layer. Then the complete encoder could be `N` deep copies of the encoder block and similarly for the complete decoder. 
I think I'm not understanding what is going on in your mind :( Could you please clarify how two inputs are flowing in the `encoderDecoderAttn` part? And how their backpropagated errors will reach to appropriate place?

Reply:
> The problem is neither we can pass more than one inputs to FFN nor to the Concat layer.

Of course we can. Technically we pass jest one input to each layer. But we can use the `Subview` layer in order to split the input in two parts. Then we'll process each input separately. Look at the example https://github.com/mlpack/mlpack/pull/2375#issuecomment-672910701.

> Could you please clarify how two inputs are flowing in the encoderDecoderAttn part?

I think the example https://github.com/mlpack/mlpack/pull/2375#issuecomment-672910701 should clarify that.

> I think defining the Backward and Gradient function manually is not a bad option (It will use less number of layers also)

Yes, the backward propagation will require less number of layers. But the implementation itself would be more complex.